File: love-brittany-temp/src/email_sender.py
"""
Email Sending Service for Relationship Reports

Handles sending HTML email reports via Gmail API.
"""

import os
import logging
import pickle
import base64
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from typing import Optional
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError

logger = logging.getLogger(__name__)

# Gmail API scopes
SCOPES = ['https://www.googleapis.com/auth/gmail.send']


class EmailSender:
    """Service for sending emails via Gmail API."""

    def __init__(self, credentials_path: str, token_path: str):
        """
        Initialize Email Sender.

        Args:
            credentials_path: Path to OAuth2 credentials JSON file
            token_path: Path to store/retrieve the token
        """
        self.credentials_path = credentials_path
        self.token_path = token_path
        self.service = None

    def authenticate(self) -> None:
        """Authenticate with Gmail API."""
        creds = None

        # Token stores the user's access and refresh tokens
        if os.path.exists(self.token_path):
            with open(self.token_path, 'rb') as token:
                creds = pickle.load(token)

        # If there are no (valid) credentials available, let the user log in
        if not creds or not creds.valid:
            if creds and creds.expired and creds.refresh_token:
                creds.refresh(Request())
            else:
                flow = InstalledAppFlow.from_client_secrets_file(
                    self.credentials_path, SCOPES)
                creds = flow.run_local_server(port=0)

            # Save the credentials for the next run
            with open(self.token_path, 'wb') as token:
                pickle.dump(creds, token)

        try:
            self.service = build('gmail', 'v1', credentials=creds)
            logger.info("Successfully authenticated with Gmail API")
        except HttpError as error:
            logger.error(f"Failed to build Gmail service: {error}")
            raise

    def send_html_email(
        self,
        to: str,
        subject: str,
        html_content: str,
        from_email: Optional[str] = None
    ) -> bool:
        """
        Send an HTML email via Gmail API.

        Args:
            to: Recipient email address
            subject: Email subject line
            html_content: HTML content of the email
            from_email: Optional sender email (defaults to authenticated user)

        Returns:
            True if email sent successfully, False otherwise
        """
        if not self.service:
            self.authenticate()

        try:
            # Create message
            message = MIMEMultipart('alternative')
            message['To'] = to
            message['Subject'] = subject

            if from_email:
                message['From'] = from_email

            # Attach HTML content
            html_part = MIMEText(html_content, 'html')
            message.attach(html_part)

            # Encode message
            raw_message = base64.urlsafe_b64encode(message.as_bytes()).decode('utf-8')

            # Send message
            send_message = {
                'raw': raw_message
            }

            result = self.service.users().messages().send(
                userId='me',
                body=send_message
            ).execute()

            logger.info(f"Email sent successfully to {to}. Message ID: {result.get('id')}")
            return True

        except HttpError as error:
            logger.error(f"Failed to send email: {error}")
            return False

    def send_plain_text_email(
        self,
        to: str,
        subject: str,
        body: str,
        from_email: Optional[str] = None
    ) -> bool:
        """
        Send a plain text email via Gmail API.

        Args:
            to: Recipient email address
            subject: Email subject line
            body: Plain text body of the email
            from_email: Optional sender email (defaults to authenticated user)

        Returns:
            True if email sent successfully, False otherwise
        """
        if not self.service:
            self.authenticate()

        try:
            # Create message
            message = MIMEText(body)
            message['To'] = to
            message['Subject'] = subject

            if from_email:
                message['From'] = from_email

            # Encode message
            raw_message = base64.urlsafe_b64encode(message.as_bytes()).decode('utf-8')

            # Send message
            send_message = {
                'raw': raw_message
            }

            result = self.service.users().messages().send(
                userId='me',
                body=send_message
            ).execute()

            logger.info(f"Email sent successfully to {to}. Message ID: {result.get('id')}")
            return True

        except HttpError as error:
            logger.error(f"Failed to send email: {error}")
            return False

    def validate_credentials(self) -> bool:
        """
        Validate Gmail API credentials.

        Returns:
            True if credentials are valid, False otherwise
        """
        if not self.service:
            try:
                self.authenticate()
            except Exception as e:
                logger.error(f"Failed to authenticate: {e}")
                return False

        try:
            # Try to get user profile
            profile = self.service.users().getProfile(userId='me').execute()
            email = profile.get('emailAddress')
            logger.info(f"Successfully validated Gmail access for: {email}")
            return True

        except HttpError as error:
            logger.error(f"Failed to validate credentials: {error}")
            return False

File: apps/todoist-daily-reviewer/src/email-sender.js
/**
 * Email Sender - Sends HTML reports via Gmail API
 *
 * Handles Gmail API authentication and email delivery.
 */

import { google } from 'googleapis';
import fs from 'fs';
import path from 'path';

/**
 * Email Sender class
 */
export class EmailSender {
  constructor(options = {}) {
    this.credentialsPath = options.credentialsPath || process.env.GOOGLE_CREDENTIALS_FILE;
    this.tokenPath = options.tokenPath || process.env.GOOGLE_TOKEN_FILE;
    this.gmail = null;
    this.initialized = false;
  }

  /**
   * Initialize Gmail API client
   */
  async initialize() {
    if (this.initialized) return;

    try {
      // Load credentials
      const credentials = await this.loadCredentials();

      // Create OAuth2 client
      const { client_secret, client_id, redirect_uris } = credentials.installed || credentials.web;
      const oAuth2Client = new google.auth.OAuth2(
        client_id,
        client_secret,
        redirect_uris ? redirect_uris[0] : 'http://localhost'
      );

      // Load saved token
      const token = await this.loadToken();
      oAuth2Client.setCredentials(token);

      // Create Gmail client
      this.gmail = google.gmail({ version: 'v1', auth: oAuth2Client });
      this.initialized = true;

      console.log('Gmail API initialized successfully');
    } catch (error) {
      console.error('Failed to initialize Gmail API:', error.message);
      throw error;
    }
  }

  /**
   * Load OAuth credentials
   */
  async loadCredentials() {
    // Check for base64 encoded credentials (Lambda environment)
    if (process.env.GMAIL_OAUTH_CREDENTIALS) {
      const decoded = Buffer.from(process.env.GMAIL_OAUTH_CREDENTIALS, 'base64').toString('utf-8');
      return JSON.parse(decoded);
    }

    // Load from file
    if (!this.credentialsPath) {
      throw new Error('No credentials path specified');
    }

    const content = fs.readFileSync(this.credentialsPath, 'utf-8');
    return JSON.parse(content);
  }

  /**
   * Load saved OAuth token
   */
  async loadToken() {
    // Check for base64 encoded token (Lambda environment)
    if (process.env.GMAIL_CREDENTIALS) {
      const decoded = Buffer.from(process.env.GMAIL_CREDENTIALS, 'base64').toString('utf-8');
      return JSON.parse(decoded);
    }

    // Load from file
    if (!this.tokenPath) {
      throw new Error('No token path specified');
    }

    const content = fs.readFileSync(this.tokenPath, 'utf-8');
    return JSON.parse(content);
  }

  /**
   * Send HTML email
   */
  async sendHtmlEmail(options) {
    const { to, subject, htmlContent, from } = options;

    if (!this.initialized) {
      await this.initialize();
    }

    // Build email
    const email = this.buildEmail({
      to,
      subject,
      htmlContent,
      from: from || 'me'
    });

    // Encode to base64url
    const encodedEmail = Buffer.from(email)
      .toString('base64')
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=+$/, '');

    try {
      const response = await this.gmail.users.messages.send({
        userId: 'me',
        requestBody: {
          raw: encodedEmail
        }
      });

      console.log(`Email sent successfully. Message ID: ${response.data.id}`);
      return {
        success: true,
        messageId: response.data.id,
        threadId: response.data.threadId
      };
    } catch (error) {
      console.error('Failed to send email:', error.message);
      throw error;
    }
  }

  /**
   * Build MIME email message
   */
  buildEmail({ to, subject, htmlContent, from }) {
    const boundary = 'boundary_' + Date.now().toString(16);

    const emailLines = [
      `From: ${from}`,
      `To: ${to}`,
      `Subject: ${subject}`,
      'MIME-Version: 1.0',
      `Content-Type: multipart/alternative; boundary="${boundary}"`,
      '',
      `--${boundary}`,
      'Content-Type: text/plain; charset="UTF-8"',
      'Content-Transfer-Encoding: quoted-printable',
      '',
      this.htmlToPlainText(htmlContent),
      '',
      `--${boundary}`,
      'Content-Type: text/html; charset="UTF-8"',
      'Content-Transfer-Encoding: quoted-printable',
      '',
      htmlContent,
      '',
      `--${boundary}--`
    ];

    return emailLines.join('\r\n');
  }

  /**
   * Convert HTML to plain text (basic conversion)
   */
  htmlToPlainText(html) {
    return html
      // Remove style and script tags with their contents
      .replace(/<style[^>]*>[\s\S]*?<\/style>/gi, '')
      .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '')
      // Replace common elements
      .replace(/<br\s*\/?>/gi, '\n')
      .replace(/<\/p>/gi, '\n\n')
      .replace(/<\/div>/gi, '\n')
      .replace(/<\/h[1-6]>/gi, '\n\n')
      .replace(/<li>/gi, '  - ')
      .replace(/<\/li>/gi, '\n')
      // Remove all remaining HTML tags
      .replace(/<[^>]+>/g, '')
      // Decode HTML entities
      .replace(/&nbsp;/g, ' ')
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"')
      .replace(/&#039;/g, "'")
      // Clean up whitespace
      .replace(/\n{3,}/g, '\n\n')
      .trim();
  }

  /**
   * Validate credentials without sending
   */
  async validateCredentials() {
    try {
      await this.initialize();

      // Try to get user profile to verify credentials work
      const response = await this.gmail.users.getProfile({ userId: 'me' });

      return {
        valid: true,
        email: response.data.emailAddress,
        messagesTotal: response.data.messagesTotal
      };
    } catch (error) {
      return {
        valid: false,
        error: error.message
      };
    }
  }
}

/**
 * Simple email sender for Lambda (without file dependencies)
 */
export class LambdaEmailSender {
  constructor() {
    this.gmail = null;
  }

  /**
   * Initialize with credentials from environment
   */
  async initialize() {
    if (this.gmail) return;

    // Get credentials from environment
    const oauthCreds = JSON.parse(
      Buffer.from(process.env.GMAIL_OAUTH_CREDENTIALS, 'base64').toString('utf-8')
    );
    const userCreds = JSON.parse(
      Buffer.from(process.env.GMAIL_CREDENTIALS, 'base64').toString('utf-8')
    );

    // Create OAuth2 client
    const { client_secret, client_id, redirect_uris } = oauthCreds.installed || oauthCreds.web;
    const oAuth2Client = new google.auth.OAuth2(
      client_id,
      client_secret,
      redirect_uris ? redirect_uris[0] : 'http://localhost'
    );

    oAuth2Client.setCredentials(userCreds);
    this.gmail = google.gmail({ version: 'v1', auth: oAuth2Client });
  }

  /**
   * Send HTML email
   */
  async send(to, subject, htmlContent) {
    await this.initialize();

    const boundary = 'boundary_' + Date.now().toString(16);
    const email = [
      `To: ${to}`,
      `Subject: ${subject}`,
      'MIME-Version: 1.0',
      `Content-Type: text/html; charset="UTF-8"`,
      '',
      htmlContent
    ].join('\r\n');

    const encodedEmail = Buffer.from(email)
      .toString('base64')
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=+$/, '');

    const response = await this.gmail.users.messages.send({
      userId: 'me',
      requestBody: { raw: encodedEmail }
    });

    return response.data;
  }
}

export default EmailSender;
